apps/infra-benchmarks/src/discriminated-union.ts

```diff
@@ -1,22 +1,189 @@
+/*
+ * This benchmark is used to find the fastest way to discriminate
+ * types of objects that are hidden under a symbol. This information
+ * will be used to best discriminate TypeGPU resources.
+ */
+
 import { Bench } from 'tinybench';
 
 const bench = new Bench({
   name: 'discriminated union',
   time: 100,
   async setup() {
-    // biome-ignore lint/suspicious/noExplicitAny: <making sure GC has no impact on the results>
+    // biome-ignore lint/suspicious/noExplicitAny: making sure GC has no impact on the results
     (globalThis as any).gc();
   },
 });
 
+const $internal = Symbol('Internal functionality');
+
+// biome-ignore format: Long array
+const STRING_TAGS = ['aaa', 'bbb', 'ccc'] as const;
+const NUMBER_TAGS = [0, 1, 2] as const;
+const NUMBER_TAG_CATALOG = {
+  aaa: 0,
+  bbb: 1,
+  ccc: 2,
+} as const;
+
+const aaa = Symbol('aaa symbol');
+const bbb = Symbol('bbb symbol');
+const ccc = Symbol('ccc symbol');
+const SYMBOL_TAGS = [aaa, bbb, ccc] as const;
+
+const NUMBER_OF_OBJS = 1000;
+
+let stringTaggedObjs: {
+  [$internal]: { type: (typeof STRING_TAGS)[number] };
+}[];
+let numberTaggedObjs: { [$internal]: { type: (typeof NUMBER_TAGS)[number] } }[];
+let symbolTaggedObjs: { [$internal]: { type: (typeof SYMBOL_TAGS)[number] } }[];
+let symbolKeyedObjs: { [K in (typeof SYMBOL_TAGS)[number]]: boolean }[];
+
 bench
-  .add('string tags', () => {
-    console.log('I am faster');
-  })
-  .add('slower task', async () => {
-    await new Promise((resolve) => setTimeout(resolve, 1)); // we wait 1ms :)
-    console.log('I am slower');
-  });
+  .add(
+    'string tags',
+    () => {
+      let count = 0;
+
+      for (const obj of stringTaggedObjs) {
+        if (obj[$internal].type === 'aaa') {
+          count += 2;
+        } else if (obj[$internal].type === 'bbb') {
+          count += 3;
+        } else if (obj[$internal].type === 'ccc') {
+          count += 4;
+        }
+      }
+
+      console.log(count);
```

Comment:
let's remove all of the console.logs, please

Reply:
sure, but it affects the result in that we count in the IO operation, which is slow and unpredictable

Reply:
from what I see, with logging the results are 2x slower, but their respective placements seem to be unaffected. so I guess it doesn't matter either way